### PR TITLE
SYCL Rewrite of reduce_by_segment

### PR DESCRIPTION
**Pull request by mmichel11**
This PR is the second half of the `by_segment` rewrites and accompanies the segmented scans in PR #608. The performance degradations for large input sizes in the original rewrite implementation have been addressed and the new implementation shows considerable speedup over what exists in main. Please contact me if you would like to have performance results. 

In addition to the rewrite, test coverage has been increased for `reduce_by_segment` in order to properly ensure functionality in certain edge cases.

This PR is dependent on the `wg_segmented_scan` implementation in PR #608 as some shared functionality exists between the two.

### CODE REVIEW

**Comment by mmichel11** [comment]
The comments that have been made on PR #608 that are applicable to this PR have been applied. The approach in the new `reduce_by_segment` implementation is identical to the scan by segments with the only difference being that we add 2 additional kernels to compute output indices for each segment's reduction.

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h [hunk 1]
```diff
@@ -138,6 +167,414 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
...
+
+template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3, typename _Range4,
+          typename _BinaryPredicate, typename _BinaryOperator>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy,
+                                                             oneapi::dpl::__internal::__difference_t<_Range3>>
+sycl_reduce_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values, _Range3&& __out_keys,
+                       _Range4&& __out_values, _BinaryPredicate __binary_pred, _BinaryOperator __binary_op,
+                       ::std::false_type /* has_known_identity */)
+{
+    typedef uint64_t CountType;
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/internal/reduce_by_segment_impl.h:200]: 1. just for keep one style - we use `using `instead of `typedef`.
2. uint64_t may not match to `neapi::dpl::__internal::__difference_t<_Range3>`
3. Actually, is better to keep just one line `return oneapi::dpl::experimental::ranges::reduce_by_segment.....` instead of lines  200-204.

**Review by MikeDvorskiy — COMMENTED** [review]
overall looks good, excepting a few minor comments.

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h [hunk 2]
```diff
@@ -138,6 +167,414 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
...
+                            // the last item must write the last index's position to return
+                            if (__i == __n - 1)
+                                __end_idx_acc[0] = __idx;
+                        }
+                    }
+                });
+        })
+        .wait();
+
+    return __end_idx.get_host_access()[0] + 1;
```
[the remark this answers is not in the document]
Comment (mmichel11) [reply include/oneapi/dpl/internal/reduce_by_segment_impl.h:575]: To clarify, should I go ahead and make this function return a future to the end index accessor? If so, I will make this change and the corresponding change to the legacy `__pattern_reduce_by_segment` which is used for types without identities in `algorithm_ranges_impl_hetero.h`

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h (hunk 1 above)
Comment (mmichel11) [reply include/oneapi/dpl/internal/reduce_by_segment_impl.h]: I have combined lines 200-204 to a single return.

File: test/support/reduce_serial_impl.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
+// -*- C++ -*-
```
[the remark this answers is not in the document]
Comment (mmichel11) [reply test/support/reduce_serial_impl.h:1]: I have made the header comment change.

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h [hunk 4]
```diff
@@ -38,6 +38,35 @@ namespace dpl
 namespace internal
 {
 
+#if _ONEDPL_BACKEND_SYCL
+// TODO: Remove prior to merge. This function is intended to live in scan_by_segment_impl.h which is not currently in the main branch
+// and is instead temporarily placed here to allow for development in a branch separate from mmichel11/by_segment_sycl_rewrites. Once PR #608
+// is merged, this can be removed
+template <typename _NdItem, typename _LocalAcc, typename _IdxType, typename _ValueType, typename _BinaryOp>
+inline _ValueType
+wg_segmented_scan(_NdItem __item, _LocalAcc __local_acc, _IdxType __local_id, _IdxType __delta_local_id,
```
[the remark this answers is not in the document]
Comment (mmichel11) [reply include/oneapi/dpl/internal/reduce_by_segment_impl.h]: Done

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h [hunk 5]
```diff
@@ -138,6 +167,414 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
...
+using _KernelName3 = __sycl_segmented_reduce_kernel3<_Name...>;
+template <typename... _Name>
+using _KernelName4 = __sycl_segmented_reduce_kernel4<_Name...>;
+} // namespace
+
+template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3, typename _Range4,
+          typename _BinaryPredicate, typename _BinaryOperator>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy,
+                                                             oneapi::dpl::__internal::__difference_t<_Range3>>
+sycl_reduce_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values, _Range3&& __out_keys,
```
[the remark this answers is not in the document]
Comment (mmichel11) [reply include/oneapi/dpl/internal/reduce_by_segment_impl.h]: Done

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h [hunk 6]
```diff
@@ -138,6 +167,414 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
...
+        ::std::forward<_ExecutionPolicy>(__exec), ::std::forward<_Range1>(__keys), ::std::forward<_Range2>(__values),
+        ::std::forward<_Range3>(__out_keys), ::std::forward<_Range4>(__out_values), __binary_pred, __binary_op);
+    return N;
+}
+
+template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3, typename _Range4,
+          typename _BinaryPredicate, typename _BinaryOperator>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy,
+                                                             oneapi::dpl::__internal::__difference_t<_Range3>>
+sycl_reduce_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values, _Range3&& __out_keys,
```
[the remark this answers is not in the document]
Comment (mmichel11) [reply include/oneapi/dpl/internal/reduce_by_segment_impl.h]: Done

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h [hunk 7]
```diff
@@ -174,10 +611,15 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
...
 
+    using has_known_identity =
+        typename __dpl_sycl::__has_known_identity<BinaryOperator,
+                                                  typename ::std::iterator_traits<InputIterator2>::value_type>::type;
+
     // number of unique keys
-    CountType N = oneapi::dpl::experimental::ranges::reduce_by_segment(
-        ::std::forward<Policy>(policy), key_buf.all_view(), value_buf.all_view(), key_output_buf.all_view(),
-        value_output_buf.all_view(), binary_pred, binary_op);
+    CountType N = sycl_reduce_by_segment(::std::forward<Policy>(policy), key_buf.all_view(), value_buf.all_view(),
```
[the remark this answers is not in the document]
Comment (mmichel11) [reply include/oneapi/dpl/internal/reduce_by_segment_impl.h]: Done

File: test/support/reduce_serial_impl.h [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                         BinaryPredicate binary_pred, BinaryOperation binary_op)
+{
+    if (n < 1)
+        return 0;
+    
+    using ValT = typename ::std::decay<decltype(vals[0])>::type;
+    using KeyT = typename ::std::decay<decltype(keys[0])>::type;
+    KeyT tmp_key = keys[0];
+    ValT tmp_val = vals[0];
+    std::size_t segment_count = 0;
```
[the remark this answers is not in the document]
Comment (mmichel11) [reply test/support/reduce_serial_impl.h]: I agree, I have made this change.

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmichel11) [reply include/oneapi/dpl/internal/reduce_by_segment_impl.h:596]: After second thought and discussion with @danhoeflinger, would it be better to add this as an issue to address after this is merged? The current mainline implementations of `__pattern_reduce_by_segment` and `*_scan_by_segment` are synchronous. I think it might be cleaner to directly address this in a separate PR especially due to the current size of this PR and time constraints.

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h [hunk 9]
```diff
@@ -1,19 +1,42 @@
-/*
- *  Copyright (c) Intel Corporation
+/*  Copyright (c) Intel Corporation
  *
- *  Copyright 2008-2013 NVIDIA Corporation
+ *  Copyright 2008-2013 NVIDIA Corporation
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/internal/reduce_by_segment_impl.h:3]: That file header is correct?
So we really use NVIDIA code in that file?

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (MikeDvorskiy) [reply include/oneapi/dpl/internal/reduce_by_segment_impl.h:596]: So, please create a "Reference in new issues"  - from "points" menu, upper right corner.

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h (hunk 9 above)
Comment (MikeDvorskiy) [reply include/oneapi/dpl/internal/reduce_by_segment_impl.h:3]: Even we borrowed some idea, we should comment it in the code with line numbers, I think..
It will help us to make a decision to change file header in the future, when the code will be modified totally, probably.

**Review by MikeDvorskiy — APPROVED** [review]
LGTM with minor comments

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h (hunk 9 above)
Comment (mmichel11) [reply include/oneapi/dpl/internal/reduce_by_segment_impl.h:3]: We do not directly use NVIDIA code but enough code was looked at for `wg_segmented_scan` that after discussion with others it was deemed necessary to provide an attribution.

I agree that we should comment in the line numbers as there are only 10 or so lines that apply. I will add this.

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmichel11) [reply include/oneapi/dpl/internal/reduce_by_segment_impl.h:596]: Created this issue.

File: test/support/reduce_serial_impl.h [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+#ifndef _REDUCE_SERIAL_IMPL_H
+#define _REDUCE_SERIAL_IMPL_H
+
+#include <iterator>
+
+template <typename ViewKeys, typename ViewVals, typename ResKeys, typename ResVals, typename Size,
+          typename BinaryPredicate, typename BinaryOperation>
+Size
+reduce_by_segment_serial(ViewKeys keys, ViewVals vals, ResKeys& res_keys, ResVals& res_vals, Size n,
```
Comment (MikeDvorskiy) [inline test/support/reduce_serial_impl.h:18]: `ViewKeys keys, ViewVals vals` are passed by value...  What type of passed parameter? An iterator or std::vector?
I see that res_keys and res_vals - `std::vector`...
I would suggest to keep consistent here  - f.e. to pass the iterators for input and output sequences.

Comment (mmichel11) [reply test/support/reduce_serial_impl.h]: `keys` and `vals` are iterators. I have made the change to pass `res_keys` and `res_vals` by value as iterators and renamed the iterator types accordingly.

**Review by MikeDvorskiy — DISMISSED** [review]
LGTM

**Review by timmiesmith — APPROVED** [review]
Reapproving after confirming all final comments by @MikeDvorskiy have been addressed.
